src/input/json_to_storage.cc

```diff
@@ -22,46 +22,184 @@ using namespace internal;
 
 
 /********************************************
- * Implementation of  internal::JSONPath
+ * Implementation of PathBase
  */
+PathBase::PathBase() {
+	path_.push_back( make_pair( (int)(-1), string("/") ) );
+}
 
 
-JSONPath::JSONPath(const Node& root_node)
+void PathBase::output(ostream &stream) const {
+    if (level() == 0) {
+        stream << "/";
+        return;
+    }
+
+    for(vector< pair<int, string> >::const_iterator it = path_.begin()+1; it != path_.end(); ++it) {
+        if ( it->first < 0 ) {
+            stream << "/" << it->second;
+        } else {
+            stream << "/" << it->first;
+        }
+    }
+}
+
+
+
+void PathBase::put_address() {
+	previous_references_.insert(str());
+}
+
+
+
+string PathBase::str() {
+    stringstream ss;
+    output(ss);
+    return ss.str();
+}
+
+
+
+/**
+ * This returns path to reference given by address in ref_address.
+ *
+ */
+PathBase * PathBase::find_ref_node(const string& ref_address)
```

Comment:
Since we use YAML references we need this method only in JSON reader so possibly we can move this method as well as previous_references_ stuff into PathJSON.